Source/Dafny/Verifier/Translator.TrStatement.cs

```diff
@@ -1910,10 +1910,14 @@ public partial class Translator {
         loopBodyBuilder.Add(Bpl.Cmd.SimpleAssign(s.Tok, w, Bpl.Expr.False));
       }
       // Finally, assume the well-formedness of the invariant (which has been checked once and for all above), so that the check
-      // of invariant-maintenance can use the appropriate canCall predicates.
+      // of invariant-maintenance can use the appropriate canCall predicates. Note, it is important (see Test/git-issues/git-issue-1812.dfy)
+      // that each CanCall assumption uses the preceding invariants as antecedents--this is achieved by treating all "invariant"
+      // declarations as one big conjunction, because then CanCallAssumption will add the needed antecedents.
+      Expression allInvariants = Expression.CreateBoolLiteral(s.Tok, true);
       foreach (AttributedExpression loopInv in s.Invariants) {
-        loopBodyBuilder.Add(TrAssumeCmd(loopInv.E.tok, CanCallAssumption(loopInv.E, etran)));
+        allInvariants = Expression.CreateAnd(allInvariants, loopInv.E);
       }
```

Comment:
This might be marginally more readable as (untested)

```
s.Invariants.Select(inv => inv.E).Aggregate(Expression.CreateAnd, Expression.CreateBoolLiteral(s.Tok, true))
```

Although I'd further simplify it by changing it to:

```
if (invariants.Any()) {
  var allInvariants = s.Invariants.Select(inv => inv.E).Aggregate(Expression.CreateAnd);
  loopBodyBuilder.Add(…);
}
```

to void the extra `true` at the end.

Reply:
Thanks. I followed your suggestion. (Annoyingly, `CreateAnd` has a third parameter, with a default value. So, I had to eta convert the argument to `Aggregate`. Still, the result is nicer than the previous loop.)